experiments/process_sandbox/library_runner.cc

```diff
@@ -239,184 +219,531 @@ namespace sandbox
         {
-          // FIXME: Report error in some useful way.
-          printf("Exception!\n");
+          uintptr_t result = call(fn);
+          if (result < 0)
+          {
+            c.set_error_return(-result);
+          }
+          else
+          {
+            c.set_success_return(result);
+          }
+          return;
         }
-        shared_mem->signal(false);
-      }
+      };
+      syscall_upcall(SyscallFrame::Open, upcall_open);
+      syscall_upcall(SyscallFrame::OpenAt, upcall_openat);
+      syscall_upcall(SyscallFrame::Stat, upcall_stat);
     }
-  };
+  }
+
 }
 
-char* ExportedLibrary::type_encoding(int idx)
+/**
+ * POSIX `open` function, performs an upcall to the host rather than a system
+ * call.
+ */
+extern "C" int open(const char* pathname, int flags, ...)
 {
-  return functions.at(idx)->type_encoding();
+  mode_t mode = 0;
+  if ((flags & O_CREAT) == O_CREAT)
+  {
+    va_list ap;
+    va_start(ap, flags);
+    mode = va_arg(ap, int);
+    va_end(ap);
+  }
+  return syscall_return(upcall_open(pathname, flags, mode));
 }
 
-int main(int, char** argv)
+#ifndef USE_CAPSICUM
+/**
+ * POSIX `openat` function, performs an upcall to the host rather than a system
+ * call.  In a Capsicum world, this is safe to allow the untrusted process to
+ * do directly, so we don't bother interposing here.
+ *
+ * TODO: We should still interpose on this with Capsicum to handle things like
+ * `openat(-100, "some/path", O_WHATEVER)`
+ */
+extern "C" int openat(int dirfd, const char* pathname, int flags, ...)
 {
-#ifdef USE_CAPSICUM
-  cap_enter();
+  va_list ap;
+  va_start(ap, flags);
+  mode_t mode = va_arg(ap, int);
+  va_end(ap);
+  int ret = upcall_openat(dirfd, pathname, flags, mode);
+  if (ret < 0)
+  {
+    errno = -ret;
+    return -1;
+  }
+  return ret;
+}
 #endif
-  void* addr = (void*)strtoull(argv[1], nullptr, 0);
-  size_t length = strtoull(argv[2], nullptr, 0);
-  // fprintf(stderr, "Child starting\n");
-  // printf(
-  //"Child trying to map fd %d at addr %p (0x%zx)\n", SharedMemRegion, addr,
-  // length);
-  void* ptr = mmap(
-    addr,
-    length,
-    PROT_READ | PROT_WRITE,
-    MAP_FIXED | MAP_ALIGNED(35) | MAP_SHARED | MAP_NOCORE,
-    SharedMemRegion,
-    0);
-  // printf("%p\n", ptr);
-  if (ptr == MAP_FAILED)
-  {
-    err(1, "Mapping shared heap failed");
+
+/**
+ * Exported function to allow the loaded code to invoke a callback to the
+ * parent.
+ */
+int sandbox::invoke_user_callback(int idx, void* data, size_t size, int fd)
+{
+  unique_c_ptr<void> copy;
+  if (
+    (data < shared_memory_start) ||
+    ((static_cast<char*>(data) + size) >= shared_memory_end))
+  {
+    copy.reset(malloc(size));
+    memcpy(copy.get(), data, size);
+    data = copy.get();
+  }
+  auto ret = upcall(static_cast<sandbox::UpcallKind>(idx), data, size, fd);
+  int result = static_cast<int>(ret.first);
+  if (ret.second.is_valid())
+  {
+    result = ret.second.take();
   }
+  return result;
+}
+
+int main()
```

Comment:
I'm guessing this would be a per-region "main" function, which will setup the sandbox and start the busy loop for the region and then tear down once the region is out of scope. The main function should only be called once the behaviour is scheduled, on the processor/core that the cown is executing. I'm assuming that apart from IPC handling, the core will be either executing parent or sandbox code, but not both at the same time (ie. no threading inside the sandbox).

The compiler should probably bundle all functionality of each region into a single object file, but we need to give them unique identifiers as to avoid clashing the static vtable in `ExternalLibrary` accesses when having multiple sandbox regions on the same binary (unless we want to share a single vtable, in which case we need to make sure two sandbox codes don't compile at the same time).

Reply:
That's not what I meant. By "all functionality of each region" I meant each invocation of a module inside a behaviour. So, if you use the same module in two different sandboxes, they'd be two different object files that get loaded when the behaviour is scheduled. Which is what you said.

My point here, however, was about the name of the function `main`.

IIUC, this is the main file per sandbox, that will set the heap, load the sandboxed library and start the busy loop, per sandbox, waiting for requests from the child process. I'm thinking if this will be part of the Verona program, we may have multiple "sandbox parent processes" running, so this `main` function would be something we call from the behaviour once it starts running (to communicate with the sandbox child process). Is that correct?

If not, then would we always fork a new parent (ie. a `library_runner`) process, per behaviour?

Reply:
I think I'm still a bit confused by your question.  In Verona, code is arranged in modules, which are classes that contain nested classes and act as a unit of encapsulation / namespacing.  A module that is implemented using foreign code is defined by a header file that describes its interface and library file (.so, .a, .dll + .lib, wasm assembly, whatever) that defines its implementation.  When you instantiate it in Verona, at the abstract-machine level, you get an `iso` object that has fields for all of the global symbols defined in that interface, which you can then use to call functions / create objects / whatever.  With the process-based sandbox, each one of those has an instance of the `SandboxedLibrary` class to define the runtime's interface to it, a shared memory region for the region's memory allocation (its heap, on the C++ side) and a running instance of the `library_runner` program, which loads the library and waits for calls.

The function here is called `main` because that is the name of the entry point for C++ programs.  The `library_runner` program is a C++ program that implements the in-sandbox run loop for a single Verona sandboxed region that is isolated using process-based sandboxing.

We will run one child process per process-sandboxed region that is live in the Verona program.  Every one of those child processes will be running the `library_runner` program and will load precisely one .so file generated by the Verona compiler (which, in turn, will link one or more .a or .so files that provide the C/C++/assembly/Rust/Intercal/whatever code that we are exposing to Verona).